ethers-etherscan/src/lib.rs

```diff
@@ -0,0 +1,454 @@
+            .join("\n")
+    }
+}
+
+/// Etherscan contract metadata
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Metadata {
+    #[serde(rename = "SourceCode")]
+    pub source_code: String,
+    #[serde(rename = "ABI")]
+    pub abi: String,
+    #[serde(rename = "ContractName")]
+    pub contract_name: String,
+    #[serde(rename = "CompilerVersion")]
+    pub compiler_version: String,
+    #[serde(rename = "OptimizationUsed")]
+    pub optimization_used: String,
+    #[serde(rename = "Runs")]
+    pub runs: String,
+    #[serde(rename = "ConstructorArguments")]
+    pub constructor_arguments: String,
+    #[serde(rename = "EVMVersion")]
+    pub evm_version: String,
+    #[serde(rename = "Library")]
+    pub library: String,
+    #[serde(rename = "LicenseType")]
+    pub license_type: String,
+    #[serde(rename = "Proxy")]
+    pub proxy: String,
+    #[serde(rename = "Implementation")]
+    pub implementation: String,
+    #[serde(rename = "SwarmSource")]
+    pub swarm_source: String,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn api_key() -> String {
+        std::env::var("ETHERSCAN_API_KEY").expect("ETHERSCAN_API_KEY not found")
+    }
+
+    #[tokio::test]
+    #[ignore]
+    async fn can_fetch_contract_abi() {
+        let api = api_key();
+        let client = Client::new("mainnet", api).unwrap();
+
+        let _abi = client
+            .contract_abi(
+                "0xBB9bc244D798123fDe783fCc1C72d3Bb8C189413"
+                    .parse()
+                    .unwrap(),
+            )
+            .await
+            .unwrap();
+    }
+
+    #[tokio::test]
+    #[ignore]
+    async fn can_fetch_contract_source_code() {
+        let api = api_key();
+        let client = Client::new("mainnet", api).unwrap();
+
+        let _meta = client
+            .contract_source_code(
+                "0xBB9bc244D798123fDe783fCc1C72d3Bb8C189413"
+                    .parse()
+                    .unwrap(),
+            )
+            .await
+            .unwrap();
+    }
+
+    #[tokio::test]
+    #[ignore]
+    async fn can_verify_contract() {
+        // TODO this needs further investigation
+
+        // https://etherscan.io/address/0x9e744c9115b74834c0f33f4097f40c02a9ac5c33#code
+        let contract = include_str!("../resources/UniswapExchange.sol");
+        let address = "0x9e744c9115b74834c0f33f4097f40c02a9ac5c33"
+            .parse()
+            .unwrap();
+        let compiler_version = "v0.5.17+commit.d19bba13";
+        let constructor_args = "0x000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000005f5e1000000000000000000000000000000000000000000000000000000000000000007596179537761700000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000035941590000000000000000000000000000000000000000000000000000000000";
```

Comment:
I wonder if we should move `BaseContract` to ethers-core now to encode constructor arguments in a nice way

Reply:
that'd be useful, I think we touched on abstracting functions as a trait via `abigen!` some time ago, I think it makes sense to follow up on this and move all general abstractions to core.